### PR TITLE
Inject data shaping parameters

### PR DESCRIPTION
**Pull request by brianjohannes**
This PR allows for injection of paradigms and data tanks into EegData. Paradigms determine how the data is reshaped for different paradigms, this replaces the need for and ErpData class. Data tank is a new addition which maintains all the raw and processed data for ease of saving. Funny enough, data maintenance is no longer a responsibility of EegData (lol). Instead, EegData is the controller of the Bessy python and follows something like a mediator pattern by having all objects interact with each other through it.

This is a big change, but I think it is for the better. 

Some bonus benefits:
- We no longer need to pre-allocate space in memory for epochs, greatly reducing memory usage
- EegData.setup() has been greatly simplified
- EegData.step() is now understandable (IMHO)

Some limitations / next steps:
- P300 is taking a long time to initialize (~20 s)
- SSVEP is currently not sending target_freqs to the classifier, I wasn't sure how to gracefully implement this with the new pattern, so I left it out for now.

### CODE REVIEW

**Comment by anup2ladder** [comment]
@brianjohannes : A number of the references to number of channels are being called `nchannels` in your PR, which contrasts with `n_channels` used throughout the rest of bessy. I had previously standardized all references to number of trials, channels, samples to the format `n_item`, e.g. `n_channels`, to be consistent with MNE (and with the group consensus). 

tldr: nchannels (and other nvariable cases) should be updated to n_channels (and n_variable)

PR in which the variable names were standardized: https://github.com/kirtonBCIlab/bci-essentials-python/pull/95

**Comment by brianjohannes** [comment]
Thanks @anup2ladder, I will change it!

**Comment by gregbci** [comment]
@brianjohannes, this looks pretty good.  Very nice to see EegData cut down to ~450 lines, nice!  Here's some suggestions:

1.  Make BaseParadigm an abstract class by inheriting from ABC.  This will prevent users from creating an instance of it.  Check out Messenger, GenericClassifier, MarkerSource, etc.

1.  BaseParadigm needs to declare what functions a BaseParadigm object must have, ex: get_eeg_start_and_end_times().  Use @abstractmethod to declare functions without bodies.  See Messenger.  It is also a compile error if a child object doesn't have the required functions.  

1.  I would rename BaseParadigm to Paradigm.  It sounds better: P300Paradigm _isa_ Paradigm type object.

1.  EegData could be renamed to BessyController, Model, Engine, Session or something more descriptive.  As you point out... it doesn't even _do_ the data part anymore 😆 

1.  You could make the properties of DataTank private to force use of the getters.

1.  Add some tasks to the backlog to look at the issues you listed.  Those can be fixed in separate PRs.

**Comment by gregbci** [comment]
Thanks @brianjohannes, yep the p300 examples run much faster now.

**Comment by gregbci** [comment]
Looks good, merging.
